### PR TITLE
OR-1256 - signed integer support

### PR DESCRIPTION
**Pull request by jjordan-quantum**
after installing hardhat and dependencies into dev, getting the following when I try to run a simple test: 

TypeError: test/contracts/ExampleContract.test.ts: Emit skipped
    at getOutput (/home/jay/dev/umb/canvas/node_modules/ts-node/src/index.ts:755:17)
    at Object.compile (/home/jay/dev/umb/canvas/node_modules/ts-node/src/index.ts:968:32)
    at Module.m._compile (/home/jay/dev/umb/canvas/node_modules/ts-node/src/index.ts:1056:42)
    at Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Object.require.extensions.<computed> [as .ts] (/home/jay/dev/umb/canvas/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:94:18)
    at /home/jay/dev/umb/canvas/node_modules/hardhat/node_modules/mocha/lib/mocha.js:349:36
    
I have updated the paths in the hardhat config to compile / use contracts in the src folder.
If i put a contracts dir in the base dir and use the default hardhat config the test is able to run.
This option would work, but i do not want to pollute the base dir of the project.

### CODE REVIEW

**Comment by jjordan-quantum** [comment]
Although the typical hardhat test was broken due to specifying non-default paths for the test and contract directories, mocha tests still work using mocha command and registering hardhat:

`$ mocha --require ts-node/register --require hardhat/register --timeout 10000 --ui bdd test/contracts/ExampleContract.test.ts`

The contracts need to be compiled prior to testing.  This works once, and then subsequent executions of 'hardhat compile' causes the following: [Error: EISDIR: illegal operation on a directory, read]

If the 'cache' and 'src/contracts/examples/artfifacts' directories are deleted, the compile script will work again.

File: hardhat.config.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        version: '0.6.8'
+      }
+    ]
+  },
+  paths: {
+    tests: './test/contracts',
+    sources: './src/contracts/examples',
+    artifacts: './src/contracts/examples/artifacts'
+  }
+};
```
Comment (DZariusz) [inline hardhat.config.ts:17]: nl

File: .npmignore [hunk 2]
```diff
@@ -5,3 +5,4 @@ tsconfig.json
 src
 test
 .env*
+hardhat.config.ts
```
Comment (DZariusz) [inline .npmignore:8]: why we ignoring it?

File: package.json [hunk 3]
```diff
@@ -25,10 +25,13 @@
     "copy:assets": "cpx 'src/**/*.{graphql,html,png,json,sol}' dist",
     "typeorm": "node --require ./node_modules/ts-node/register ./node_modules/typeorm/cli.js",
     "build": "npm run clean && npm run compile && npm run copy:assets",
-    "test": "npx mocha --require ts-node/register --ui bdd 'test/**/*.test.ts' --timeout=30000",
+    "test": "npx mocha --require ts-node/register --require hardhat/register  --ui bdd 'test/**/*.test.ts' --timeout=30000",
+    "test:example": "mocha --require ts-node/register --require hardhat/register --timeout 10000 --ui bdd test/contracts/ExampleContract.test.ts",
     "prepare": "npm run build"
   },
   "devDependencies": {
+    "@nomiclabs/hardhat-ethers": "^2.0.4",
```
Comment (DZariusz) [inline package.json:33]:
```suggestion
    "@nomiclabs/hardhat-ethers": "~2.0.4",
```

use ~ everywhere

File: src/contracts/lib/ValueDecoder.sol [hunk 4]
```diff
@@ -13,4 +13,23 @@ library ValueDecoder {
...
+      return i;
+    } else { // negative values
+      assembly {
+        i := sub(sub(u, max), 1)
+      }
+    }
+
+    return i;
+  }
+}
```
Comment (DZariusz) [inline src/contracts/lib/ValueDecoder.sol:35]: nl

File: .npmignore (hunk 2 above)
Comment (jjordan-quantum) [reply .npmignore]: -sorry..this was an attempt to fix an issue with tsc not liking hardhat.config.ts being outside of rootDir

-the fix was to remove it from the "files" key of tsconfig (it was only put there as a recommendation by hardhat)

-i will remove it

File: src/contracts/examples/ExampleContract.sol [hunk 5]
```diff
@@ -0,0 +1,79 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.6.8;
+pragma experimental ABIEncoderV2;
+
+//import "hardhat/console.sol";
+
+import "../lib/ValueDecoder.sol";
+
+// contract with examples for working with signed integers in solidity
```
Comment (DZariusz) [inline src/contracts/examples/ExampleContract.sol:9]: the goal is not to show how to work with signet integers, but to show how to use new added method to `ValueDecoder` in scope of our oracle.

File: src/contracts/examples/artifacts/src/contracts/examples/ExampleContract.sol/ExampleContract.dbg.json [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (DZariusz) [inline src/contracts/examples/artifacts/src/contracts/examples/ExampleContract.sol/ExampleContract.dbg.json:1]: let's add artifacts to ignore list

**Review by DZariusz — CHANGES_REQUESTED** [review]
the goal is not to show how to work with signet integers, but to show how to use new added method to `ValueDecoder` in scope of our oracle. No need to create this new contract, we already have contract with examples, all we need to do is pick the value that has type of INT (in our merke tree, so we need to use API), and then we can show how to use it on-chain.

Unfortunately this can be done only after we:
- deploy Chain with this feature (at least to dev)
- validator will be ready to support INT values

**Comment by DZariusz** [comment]
we also missing coders/decoders modification - this feature is required first, in order to add support for INT in pegasus.

File: src/contracts/examples/ExampleContract.sol (hunk 5 above)
Comment (jjordan-quantum) [reply src/contracts/examples/ExampleContract.sol:9]: sorry i misunderstood:
'add examples to solidity (in SDK) for how to use int'
- i can remove these functions from the example contract and test cases

-and the example contract can be used to provide a means to test the new ValueDecoder method?

File: src/contracts/examples/ExampleContract.sol [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    function convertUintToInt(uint224 u) external pure returns (int256) {
+        return u.toInt();
+    }
+
+    // example usage of ValueDecoder.toUint function
+    // expects a int224 in raw bytes
+    function convertIntToUInt(int224 i) external pure returns (uint224) {
+
+        // first convert signed int input to bytes
+        bytes memory b = new bytes(32);
```
Comment (avoloshko) [inline src/contracts/examples/ExampleContract.sol:25]: is it possible to encapsulate this code?

Comment (jjordan-quantum) [reply src/contracts/examples/ExampleContract.sol]: this function has been removed

**Comment by jjordan-quantum** [comment]
latest commit includes test cases for functions implemented in LeafValueCoder to code/decode signed integers when int prefix is present.  One thing I am curious about is the LeafValue type....being a number | HexStringWith0x, it limits signed ints to a range of MIN_SAFE_INT to MAX_SAFE_INT.  This range is what I've included in the test cases.

File: src/constants.ts [hunk 8]
```diff
@@ -5,3 +5,7 @@ export const CHAIN_CONTRACT_NAME_BYTES32 = '0x436861696e000000000000000000000000
 export const NUMERIC_MULTIPLIER = 10 ** 18;
 
 export const FIXED_NUMBER_PREFIX = 'FIXED_';
+
+export const INT_PREFIX = 'INT_';
+
+export const MAX_UINT224_HEX = '0x00000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
```
Comment (DZariusz) [inline src/constants.ts:11]: nl

File: src/contracts/examples/ExampleContract.sol [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+// contract with examples for working with signed integers in solidity
+contract ExampleContract {
+    using ValueDecoder for uint224;
+
+    // example usage of ValueDecoder.toInt function
+    // expects a signed integer encoded as a uint224
+    function convertUintToInt(uint224 u) external pure returns (int256) {
+        return u.toInt();
+    }
+}
```
Comment (DZariusz) [inline src/contracts/examples/ExampleContract.sol:18]: nl

File: src/constants.ts (hunk 8 above)
Comment (jjordan-quantum) [reply src/constants.ts:11]: yeah im having issues with this...the code has terminal newlines in my code editor, plus i run the formatters already defined in the package.json script, and what you see is what git is pushing

File: src/services/LeafValueCoder.ts [hunk 10]
```diff
@@ -1,8 +1,11 @@
 import BigNumber from 'bignumber.js';
-import { FIXED_NUMBER_PREFIX, NUMERIC_MULTIPLIER } from '../constants';
+import { FIXED_NUMBER_PREFIX, NUMERIC_MULTIPLIER, INT_PREFIX, MAX_UINT224_HEX } from '../constants';
 import { utils } from 'ethers';
 import { evenHex, prepend0x } from '../utils/helpers';
 import { FeedValue } from '../types/Feed';
+const maxUint224 = new BigNumber(MAX_UINT224_HEX);
```
[the remark this answers is not in the document]
Comment (DZariusz) [reply src/services/LeafValueCoder.ts:6]: I know it was BigNumber, but if we can do it with minimal work, let's slowly switch to `bigint` where we can.

[the remark this answers is not in the document]
Comment (jjordan-quantum) [reply src/services/LeafValueCoder.ts:6]: bigNumber was being used in the previous version of the LeafValueDecoder, in the decode and fixedValueToHex functions -> I will update all usage to bigInt?

[the remark this answers is not in the document]
Comment (jjordan-quantum) [reply src/services/LeafValueCoder.ts:6]: So to clarify, use `bigInt` for all new code requiring dealing with large integers, and leave existing code in LeafValueCoder to use `BigNumber`?

And are you referring to `BigInt`?

[the remark this answers is not in the document]
Comment (DZariusz) [reply src/services/LeafValueCoder.ts:6]: if this requires some deep code changes, then ignore it

File: src/services/LeafValueCoder.ts [hunk 11]
```diff
@@ -74,4 +91,22 @@ export class LeafValueCoder {
...
+
+  static intValueToHex = (n: BigNumber | number): string => {
+    if (n >= 0) {
+      return n.toString(16);
+    }
+
+    return maxUint224.plus(n).plus(1).toString(16);
+  };
+
+  static toInt = (n: BigNumber): number => {
```
Comment (DZariusz) [inline src/services/LeafValueCoder.ts:105]:
```suggestion
  static toInt = (n: BigNumber): bigint => {
```

number will not handle all required values, we will loose precision

File: src/services/LeafValueCoder.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (jjordan-quantum) [reply src/services/LeafValueCoder.ts:6]: BigNumber only used for manipulating numbers internally in LeafValueCoder class, they are not exposed outside the functions. I should be able to update these usages to BigInt and run the existing coder/decoder tests for QA

**Review by DZariusz — CHANGES_REQUESTED** [review]
left some comments, also we need few tests to be added here: https://github.com/umbrella-network/canvas/blob/main/test/integration/coders.test.ts

File: src/services/LeafValueCoder.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (DZariusz) [reply src/services/LeafValueCoder.ts:6]: > So to clarify, use `bigInt` for all new code required to deal with large integers, and leave existing code in LeafValueCoder to use `BigNumber`?
> 

yes, for new code

> And are you referring to `BigInt`?

`BigInt` is constructor name, `bigint` is a type name

File: src/services/LeafValueCoder.ts (hunk 11 above)
Comment (jjordan-quantum) [reply src/services/LeafValueCoder.ts:105]: > left some comments, also we need few tests to be added here: https://github.com/umbrella-network/canvas/blob/main/test/integration/coders.test.ts

**Comment by jjordan-quantum** [comment]
> left some comments, also we need few tests to be added here: https://github.com/umbrella-network/canvas/blob/main/test/integration/coders.test.ts

ok will do

File: src/services/LeafValueCoder.ts [hunk 12]
```diff
@@ -12,10 +15,23 @@ export class LeafValueCoder {
...
+
+    if (LeafValueCoder.isIntValue(label)) {
+      if (typeof n === 'number' || typeof n === 'bigint') {
+        return LeafValueCoder.encodeHex(LeafValueCoder.intValueToHex(n), bits);
+      } else {
+        throw Error(`${n} is not valid number or bigint`);
+      }
+    }
+
+    if (typeof n === 'number') {
```
Comment (DZariusz) [inline src/services/LeafValueCoder.ts:30]: let's not do that, it should be ok if value is passed as string, also we do not have any issues with that part of the code, let's not change behaviour

Comment (DZariusz) [reply src/services/LeafValueCoder.ts:30]: oh you changed the `FeedValue` type, below line should work for not numbers, so we should not need this condition

File: src/services/LeafValueCoder.ts [hunk 13]
```diff
@@ -71,7 +96,26 @@ export class LeafValueCoder {
     if (typeof n === 'number') {
       return n.toString(16);
     }
+    if (typeof n === 'string') {
+      return new BigNumber(n || '0', n.startsWith('0x') ? 16 : 10).toString(16);
+    }
+    throw Error(`${n} is not valid number or hex string`);
```
Comment (DZariusz) [inline src/services/LeafValueCoder.ts:102]: bigint is valid number, will line 100 not work for bigint? if it will, then we not need to. change anything here.

File: src/services/LeafValueCoder.ts [hunk 14]
```diff
@@ -71,7 +96,26 @@ export class LeafValueCoder {
...
+  };
+
+  static isIntValue = (label: string): boolean => label.startsWith(INT_PREFIX);
+
+  static intValueToHex = (n: bigint | number): string => {
+    if (n >= 0) {
+      return n.toString(16);
+    }
+    return (typeof n === 'bigint' ? maxUint224 + n + 1n : maxUint224 + BigInt(n) + 1n).toString(16);
```
Comment (DZariusz) [inline src/services/LeafValueCoder.ts:111]: nl after multiline

Comment (DZariusz) [inline src/services/LeafValueCoder.ts:111]:
```suggestion
    return (maxUint224 + BigInt(n) + 1n).toString(16);
```

File: test/integration/coders.test.ts [hunk 15]
```diff
@@ -23,4 +27,13 @@ describe('LeafValueCoder', () => {
       expect(LeafValueCoder.decode(leaf, FIXED_NUMBER_PREFIX)).to.eql(result);
     });
   });
+
+  const bigints: bigint[] = [0n, 1n, maxInt224, minInt224, -1n];
```
Comment (DZariusz) [inline test/integration/coders.test.ts:31]: let's add few cases for "ordinary" numbers

```suggestion
  const bigints: bigint[] = [minInt224, -123456789.987654321, -33.44, -1n, 0n, 1n, 11, 123.321, maxInt224];
```

**Review by DZariusz — COMMENTED** [review]
I think it looks ok, we should start testing this locally, and on dev.

I would add more tests for ordinary numbers, most of time you testing only like edge cases, max, 0, 1.

**Comment by jjordan-quantum** [comment]
> I think it looks ok, we should start testing this locally, and on dev.
> 
> I would add more tests for ordinary numbers, most of time you testing only like edge cases, max, 0, 1.

 I will make the requested changes.  Most of the type checking conditions were required to prevent compiler errors, but I will have another look.

File: src/services/LeafValueCoder.ts (hunk 12 above)
Comment (jjordan-quantum) [reply src/services/LeafValueCoder.ts:30]: the BigNumber constructor wont accept bigint as value, but i use the following and all the tests pass:
`return LeafValueCoder.encodeHex(new BigNumber(n.toString(), 10).times(NUMERIC_MULTIPLIER).toString(16), bits);`
(removed the condition on line 30)

File: src/services/LeafValueCoder.ts (hunk 13 above)
Comment (jjordan-quantum) [reply src/services/LeafValueCoder.ts:102]: BigNumber wont accept bigint as constructor arg, maybe I will try taking toString() like comment above

Comment (jjordan-quantum) [reply src/services/LeafValueCoder.ts:102]:
```suggestion
    return new BigNumber(n.toString() || '0', n.toString().startsWith('0x') ? 16 : 10).toString(16);
```
tests pass with this change, allows the removal of condition. let me know if you see an issue with this?
